src/crystal/system/unix/file_descriptor.cr

```diff
@@ -0,0 +1,145 @@
+require "c/fcntl"
+
+# :nodoc:
+module Crystal::System::FileDescriptor
+  include IO::Syscall
+
+  @fd : Int32
+
+  @read_event : Event::Event?
+  @write_event : Event::Event?
+
+  private def unbuffered_read(slice : Bytes)
+    read_syscall_helper(slice, "Error reading file") do
+      # `to_i32` is acceptable because `Slice#size` is a Int32
+      LibC.read(@fd, slice, slice.size).to_i32
+    end
+  end
+
+  private def unbuffered_write(slice : Bytes)
+    write_syscall_helper(slice, "Error writing file") do |slice|
+      LibC.write(@fd, slice, slice.size).tap do |return_code|
+        if return_code == -1 && Errno.value == Errno::EBADF
+          raise IO::Error.new "File not open for writing"
+        end
+      end
+    end
+  end
+
+  private def system_blocking?
+    fcntl(LibC::F_GETFL) & LibC::O_NONBLOCK == 0
+  end
+
+  private def system_blocking=(value)
+    current_flags = fcntl(LibC::F_GETFL)
+    new_flags = current_flags
+    if value
+      new_flags &= ~LibC::O_NONBLOCK
+    else
+      new_flags |= LibC::O_NONBLOCK
+    end
+    fcntl(LibC::F_SETFL, new_flags) unless new_flags == current_flags
+  end
+
+  private def system_close_on_exec?
+    flags = fcntl(LibC::F_GETFD)
+    (flags & LibC::FD_CLOEXEC) == LibC::FD_CLOEXEC
+  end
+
+  private def system_close_on_exec=(arg : Bool)
+    fcntl(LibC::F_SETFD, arg ? LibC::FD_CLOEXEC : 0)
+    arg
+  end
+
+  def self.fcntl(fd, cmd, arg = 0)
+    r = LibC.fcntl(fd, cmd, arg)
+    raise Errno.new("fcntl() failed") if r == -1
+    r
+  end
+
+  def fcntl(cmd, arg = 0)
```

Comment:
AFAIK: `IO#fcntl method is supposed to be public, but it won't be documented?

Reply:
It wasn't documented before, I can add documentation if you want.

Reply:
No, I meant: will the `#fcntl` method still appear in the documentation?

Reply:
Hmm, not sure. Let me check.

Reply:
They actually break the doc generator by saying they're included from the module, but the module is `:nodoc:` so that the links don't work.

Reply:
I'll go and do this method the same as the rest - with the `system_` prefix and `{% ifdef %}` the method out.

Reply:
Yes, I've noticed this, too. Inclusion and extension of modules that are not documented should probably be hidden in the docs generator.

Reply:
I pushed a fixup commit which resolves this. If this is OK i'll rebase to allow this to be merged.